aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/RegionDecorator.kt

```diff
@@ -223,3 +216,14 @@ class RegionProviderConfig(codegenContext: ClientCodegenContext) : ConfigCustomi
 }
 
 fun region(runtimeConfig: RuntimeConfig) = AwsRuntimeType.awsTypes(runtimeConfig).resolve("region")
+
+/**
+ * Test if region is used and configured for a model (and available on a service client).
+ *
+ * Services that have an endpoint ruleset that references the SDK::Region built in, or
+ * that use SigV4, both need a configurable region.
+ */
+fun usesRegion(codegenContext: ClientCodegenContext) =
+    codegenContext.getBuiltIn(AwsBuiltIns.REGION) != null ||
+        ServiceIndex.of(codegenContext.model)
+            .getEffectiveAuthSchemes(codegenContext.serviceShape).containsKey(SigV4Trait.ID)
```

Comment:
Do we ever have to worry about models supporting sigv4a but not sigv4? How do retry partitions interact with Sigv4a regions (if at all)?